Identity cleanup (#14)
tomleavy/mimi-arch-draft: draft-aegis-mimi-arch-latest.md

## X.509 Identity Provider

The X.509 identify provider allows for hierarchical identities based on
certificate chains. The provider is configured via the `X509Parameters` struct. 
Besides clients, depending on how the provider is configured it may also support
accounts and/or domains. Conceptually, an account is a collection of clients
(usually belonging to one user) and a domain is a collection of accounts
(e.g. belonging to a particular organization).

> The only credential supported by the X509 Identity Provider are `x509` 

Reviewer: ```suggestion
The only credential type supported by the X509 Identity Provider is `x509` .
```

> credentials which consist of a chain of 1 or more X.509 certificates. 

Reviewer: ```suggestion
An X.509 credential contains a chain of 1 or more X.509 certificates, encoded as specified in {{!I-D.ietf-mls-protocol}}.
```
The point is that MLS already defines these certs.

> Certificate chains MUST be validated according to the rules in RFC 5280 using

Reviewer: ```suggestion
Certificate chains MUST be validated according to the rules in {{!RFC5280}} using
```

the trust roots agreed upon by the (TODO: Some sort of extension dealing with
trust root negotiation).

~~~
struct {
    uint64 msg_timestamp;
} X509ApplicationContext;

struct {
    uint8 start;
    uint8 end<0...255>;

> } X509EndpointHandleRange

Reviewer: "Endpoint" is confusing if in the same doc we have REST endpoints... Maybe session for the lack of a better word?

struct {

> 	bool use_accounts;

Reviewer: This TLS presentation language doesn't have a bool, but maybe it doesn't matter because we never send it on the wire?

	bool use_domains:
    X509EndpointHandleRange endpoint_handle_range;
	select (use_accounts) {
	    case true:
		    uint8 account_handle_offset;

> 		case false:

Reviewer: [Nit] Tabs don't seem to render very well.

		    struct {}
	}
	select (use_domains) {
	    case true:
		    uint8 domain_name_offset;
		case false:
		    struct {}
	}
} X509Parameters;
~~~

A given parameter set `X509Parameters params` is valid if the following
conditions are all met. The end point handle range MUST not end before it
starts.

~~~
Assert (params.endpoint_handle_range.end >= params.endpoint_handle_range.start)
~~~

If the X509 Identity Provider is configured to use accounts then the account

> handel offset MUST strictly succeeds the end point handle range in the

Reviewer: ```suggestion
handel offset MUST strictly succeed the end point handle range in the
```

certificate chain counting from the leaf certificate up. 

~~~
If (params.use_accounts == true) {
    Assert (params.account_handle_offset > params.endpoint_handle_range.end)
}
~~~~

If the provider is configured to use domains then the domain name offset MUST

> strictly succeeds the end point handle range in the certificate chain counting

Reviewer: ```suggestion
strictly succeed the end point handle range in the certificate chain counting
```

from the leaf certificate up.

~~~
If (params.user_domains == true) {
    Assert (params.domain_name_offset > params.endpoint_handle_range.end)
}

> ~~~~

Reviewer: ```suggestion
~~~
```

Finally, if both domains and accounts are used then the domain name offset MUST
also strictly succeed the account handle offset.

~~~
If (params.use_domains == true) && (params.use_accounts == true) {
    Assert (params.domain_name_offset > params.account_handle_offset)
~~~~

### X.509 Identifiers

Each client has a handle (i.e. a human readable name). If the X509 Identity

> Provider is configured to not use domains then client handles are unique across

Reviewer: "are" => "MUST be"?

all clients registered in the application. However, if domains are being used
then client handles are only unique within scope of the client's domain.

If the Provider is configured to use accounts then a client handle is defined

> to be the concatenation of its account handel followed by its end point handle

Reviewer: ```suggestion
to be the concatenation of its account handle followed by its end point handle
```
Funny, "handel"" means trade in polish

seperated by a "/". When accounts are not used the client handle is simply its
end point handle.

~~~
If (params.use_accounts == true) {
    client_handle := account_handle + "/" + end_point_handle;
} Else {
    client_handle := end_point_handle;
}
~~~

End piont handles, account handles (when used) and domain names (when used) are

> derived from subject Common Name fields in a clients credential certificate

Reviewer: ```suggestion
derived from subject Common Name fields in a client's credential certificate
```

chain as described bellow. To ensure unambiguous URI's for clients and accounts
(as described in {{?I-D.draft-mahy-mimi-identity}}) the Common Name fields found

> used to derive the handles MUST not contain any of the 5 special characters

Reviewer: ```suggestion
used to derive the handles MUST NOT contain any of the 5 special characters
```
